### PR TITLE
Add support for reporting on Google spreadsheet

### PR DESCRIPTION
**Pull request by malesci**
Integrated Google Spreadsheet to add a new row for each execution of the script

### CODE REVIEW

**Comment by jjjchens235** [comment]
On my list to take a look at.

**Comment by jjjchens235** [comment]
sorry i still haven't look at this yet.. i have not forgotten.

**Comment by malesci** [comment]
> sorry i still haven't look at this yet.. i have not forgotten.

No issue, feel free to ask in case need help.

**Comment by jjjchens235** [comment]
Could you also provide a screenshot in the README so that the user knows what data will be reported on?

**Comment by malesci** [comment]
> Do you mind providing a screenshot of what this looks like Google Sheets?
> In a recent commit, I refactored `print_stats()` and I want to make sure when I'm cleaning up merge conflicts with this PR, I'm capturing the spirit of what you're after.

Sure, here the rows added on my spreadsheet.
Please could you add it on the README as well?

Schermata 2022-02-13 alle 15 52 42

regards

**Comment by jjjchens235** [comment]
Thanks for fast response.
Spreadsheet looks nice!
Yea I can add to README.md

**Comment by malesci** [comment]
hello, have some news about google spreadsheet support? Thank you

**Comment by jjjchens235** [comment]
Sorry this is my mistake, I had created a review but I never submitted my review, it was in a 'pending' state.

File: BingRewards/BingRewards.py [hunk 1]
```diff
@@ -77,7 +78,22 @@ def __main():
     else:
         telegram_messenger = TelegramMessenger(telegram_api_token, telegram_userid)
     
-    rewards = Rewards(email, password, telegram_messenger, DEBUG, args.headless, cookies, args.driver)
```
[the remark this answers is not in the document]
Comment (malesci) [reply BingRewards/BingRewards.py:80]: > 1. Where are googlespreadsheet_id / googlespreadsheet_sheetname obtained in the first place?
From the URL (e.g. https://docs.google.com/spreadsheets/d/SPREADSHEET_ID/edit#gid=0), copy the SPREADSHEET_ID (e.g. 1hlTYOqy10Q-1q32ZEkdeBx_D9y_c283_ZHLvPOcJ6b1)

sheetname is the name of your sheet, user can defined. Default one should be Sheet1

> 2. Why are we writing values to credentials.json / token.json?
> 
> Edit:
> 
> 1. Furthermore, can some logic be added to setup.py so that the necessary credential values can be inputted and then encoded by the program rather than by the user? Note: I just updated setup.py in the dev [branch](https://github.com/jjjchens235/bing-rewards/blob/dev/setup.py#L48).

yes of course it would be better to add some logic like this, honestly it's the fastest way I've found. Any improvement is always welcome :)

File: README.md [hunk 2]
```diff
@@ -75,6 +79,20 @@ if you want to setup telegram notification system please follow these steps:
 	- your token generated from step 1
 	- your userid from step 2
 
+## Google Spreadsheet reporting
```
[the remark this answers is not in the document]
Comment (malesci) [reply README.md:82]: maybe I'm wrong but I don't think there are differences at the country level, my instructions were simply a little old I think.
I don't know which parts of the json exactly you need to store, I think you need everything, that's why I hashed the whole file.

File: BingRewards/BingRewards.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jjjchens235) [reply BingRewards/BingRewards.py:80]: Thank you, could you also elaborate on this please `Why are we writing values to credentials.json / token.json?`

**Comment by jjjchens235** [comment]
There is going to be a lot of refactoring that needs to be done on this, just due to all the changes since your initial commit.

Would you like to refactor it, or would you prefer me to do it? I would pull in your commit and build on top of it. Totally your call since it's my bad that this work got delayed for so long.

**Comment by malesci** [comment]
> There is going to be a lot of refactoring that needs to be done on this, just due to all the changes since your initial commit.
> 
> Would you like to refactor it, or would you prefer me to do it? I would pull in your commit and build on top of it. Totally your call since it's my bad that this work got delayed for so long.

I’m not a developer, my job maybe will be not so good but tomorrow I will try to find some time and redo the commit from your master branch. Not able to promise 😀

**Comment by jjjchens235** [comment]
Sounds like a plan. Here are some notes I took on what needs to be done, feel free to implement differently if you have better solution.

## Outstanding questions

#### Credentials.json vs Config.py
One of the steps in the README is the following:
`Encode credentials.json text to Base64 format and store on config.py`

I don't think this is necessary, it's going to be a pain in the ass to encode each key/ value pair.

I would prefer to only store in `config.py` any values that are not included in `credentials.json` but are needed for the program, i.e spreadsheetid. These extra values should be prompted in `setup.py` so that the program can encode these values for the user.

Also, what path does `googlespreadsheet.py` expect `credentials.json` to be at, can that be included in the README?


## Refactoring
What refactoring needs to be done?

I think it needs to be done in the following places from most to least important
- in `rewards.py` , there should be no reference to the `google_sheets` object.
- In `BingRewards.py`, the `rewards.stats` object should be passed into the `google_sheets` object.
- In `setup.py`, the user should be prompted to enter any credential info missing in `credentials.json`. 
- README.md needs to be updated with better `Google API` instructions

## Refactor Rewards.py
Objects should no longer be referenced directly in `rewards.py`, i.e telegram or googlesheets objects.

Rather, this logic should be moved to `BingRewards.py` which serves as the 'controller' between the rewards object and other objects.

## BingRewards.py
Firstly, inside `BingRewards.py` The rewards object now has an attribute `stats` that can be passed into any object that needs the stats information.

Here is the [ telegram example](https://github.com/jjjchens235/bing-rewards/blob/master/BingRewards/BingRewards.py#L96)

Secondly, create a method called `get_googlesheets...()` , [telegram example](https://github.com/jjjchens235/bing-rewards/blob/master/BingRewards/BingRewards.py#L33)

## Refactor Setup.py
The user should be prompted for any information that is not stored in `credentials.json` already.

To implement this, just need to add a few lines like how we did for [telegram](https://github.com/jjjchens235/bing-rewards/blob/master/setup.py#L51-L52)

## Update readme
The instructions are hard to follow for a new user, possibly due to outdated instructions.

**Comment by malesci** [comment]
> Sounds like a plan. Here are some notes I took on what needs to be done, feel free to implement differently if you have better solution.
> 
> ## Outstanding questions
> #### Credentials.json vs Config.py
> One of the steps in the README is the following: `Encode credentials.json text to Base64 format and store on config.py`
> 
> I don't think this is necessary, it's going to be a pain in the ass to encode each key/ value pair.
> 
> I would prefer to only store in `config.py` any values that are not included in `credentials.json` but are needed for the program, i.e spreadsheetid. These extra values should be prompted in `setup.py` so that the program can encode these values for the user.
> 
> Also, what path does `googlespreadsheet.py` expect `credentials.json` to be at, can that be included in the README?
> 
> ## Refactoring
> What refactoring needs to be done?
> 
> I think it needs to be done in the following places from most to least important
> 
> * in `rewards.py` , there should be no reference to the `google_sheets` object.
> * In `BingRewards.py`, the `rewards.stats` object should be passed into the `google_sheets` object.
> * In `setup.py`, the user should be prompted to enter any credential info missing in `credentials.json`.
> * README.md needs to be updated with better `Google API` instructions
> 
> ## Refactor Rewards.py
> Objects should no longer be referenced directly in `rewards.py`, i.e telegram or googlesheets objects.
> 
> Rather, this logic should be moved to `BingRewards.py` which serves as the 'controller' between the rewards object and other objects.
> 
> ## BingRewards.py
> Firstly, inside `BingRewards.py` The rewards object now has an attribute `stats` that can be passed into any object that needs the stats information.
> 
> Here is the [ telegram example](https://github.com/jjjchens235/bing-rewards/blob/master/BingRewards/BingRewards.py#L96)
> 
> Secondly, create a method called `get_googlesheets...()` , [telegram example](https://github.com/jjjchens235/bing-rewards/blob/master/BingRewards/BingRewards.py#L33)
> 
> ## Refactor Setup.py
> The user should be prompted for any information that is not stored in `credentials.json` already.
> 
> To implement this, just need to add a few lines like how we did for [telegram](https://github.com/jjjchens235/bing-rewards/blob/master/setup.py#L51-L52)
> 
> ## Update readme
> The instructions are hard to follow for a new user, possibly due to outdated instructions.

Thanks a lot for the well structured and detailed plan.
Unfortunately I did not have time to do something but I see that you are much more on the piece than me, can I ask you to take care of these changes? Thank you so much for that.

An important comment/request about the use of setup.py. Basically I completely agree with your approach but in my configuration **I can't use setup.py as I don't have the possibility to use it in interactive mode** in my system.

This is in fact the reason why in my commit I entered the information directly in the configuration file without going through setup.py. 
For the same reason I also opened the request for improvement for the Telegram setting (#146 ).
For me it would be very appreciated if you could allow setup with setup.py in non-interactive mode.

thank you in advance

**Comment by jjjchens235** [comment]
Sounds good, I'll finish the refactoring then.

Regarding setup.py non interactive mode, I'm not sure how to add support for that. The whole point of setup.py is to allow the user to interactively add their credentials.

Could you give me an idea for what you're imagining?

**Comment by jjjchens235** [comment]
Maybe you are suggesting to allow the user to enter their info in the prompt itself, i.e
`python setup.py -u user123` ?

**Comment by malesci** [comment]
> Sounds good, I'll finish the refactoring then.

Thank you so much

> 
> Regarding setup.py non interactive mode, I'm not sure how to add support for that. The whole point of setup.py is to allow the user to interactively add their credentials.
> 
> Could you give me an idea for what you're imagining?
> 
> Edit: I re-read your initial issue and I see what you mean. Yes, I can try working on that. Maybe I will do that first.
> 

You got exactly the point.

Let me describe my scenario for completeness, basically in my configuration I use GitHub Actions to run this application on a scheduled basis through GitHub workflow. In fact in this scenario it is not good to use the setup.py in interactive mode and for this I had to devise workaround.

> Are we still in agreement that `credentials.json` will be used primarily and we are only adding to config.py any **extra** arguments such a `spreadsheet_id` ?

Totally agree on this.

**Comment by jjjchens235** [comment]
Finally had a chance to work on this today, made some decent progress, will try to finish it up over the coming weeks.

**Comment by jjjchens235** [comment]
In fact, I got it to write to `google sheets` and all the code is pretty much refactored. 

90% there - I just need to double-check all the code, then I'll create a PR for you to review.

Ideally, there would be a way to write the `field names` too, if you have any idea how to do that, lmk!

**Comment by jjjchens235** [comment]
yea column header

**Comment by malesci** [comment]
> yea column header

Not tested by I think you can insert/update values on fixed row A1, something like this I guess

my_list = [['col1', 'col2', 'col3', 'col..n']]

worksheet.values_update(
    'Sheet1!A1',
    params={
        'valueInputOption': 'RAW'
    },
    body={
        'values':my_list
    }
)

ref. https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets.values/update

**Comment by jjjchens235** [comment]
Thanks for your instructions for column header, that was helpful.

I have pushed all the changes to the `dev` branch, do you mind testing the changes to make sure they work for you?

**Comment by malesci** [comment]
> I have pushed all the changes to the `dev` branch, do you mind testing the changes to make sure they work for you?

very great job, thank you so much. 
I will do a check in the next coming days.

**Comment by malesci** [comment]
> Thanks for your instructions for column header, that was helpful.
> 
> I have pushed all the changes to the `dev` branch, do you mind testing the changes to make sure they work for you?

It works like a charm. Thank you

**Comment by jjjchens235** [comment]
Thanks for confirming. It's in master now, thanks for your patience on this one.

As an FYI, I ran into this small issue here, currently the program prints out a manual resolution. Lmk if you come up with a better way to handle:
https://stackoverflow.com/questions/66895957/google-api-with-python-error-when-trying-to-refresh-token
